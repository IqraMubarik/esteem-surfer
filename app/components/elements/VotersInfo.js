import React, { Component } from 'react';

import PropTypes from 'prop-types';

import { FormattedMessage, FormattedRelative } from 'react-intl';
import { Popover, Modal, Table, Badge } from 'antd';

import parseToken from '../../utils/parse-token';
import currencySymbol from '../../utils/currency-symbol';
import parseDate from '../../utils/parse-date';
import authorReputation from '../../utils/author-reputation';

export const prepareContentVotes = (entry, currencyRate) => {
  const totalPayout =
    parseToken(entry.pending_payout_value) +
    parseToken(entry.total_payout_value) +
    parseToken(entry.curator_payout_value);

  const voteRshares = entry.active_votes.reduce(
    (a, b) => a + parseFloat(b.rshares),
    0
  );
  const ratio = totalPayout / voteRshares;

  return entry.active_votes
    .map(a => {
      const rew = a.rshares * ratio * currencyRate;

      return Object.assign({}, a, {
        reputation: authorReputation(a.reputation),
        reward: rew,
        time: parseDate(a.time),
        percent: a.percent / 100
      });
    })
    .sort((a, b) => {
      const keyA = a.reward;
      const keyB = b.reward;

      if (keyA > keyB) return -1;
      if (keyA < keyB) return 1;
      return 0;
    });
};

export const preparePopoverContent = (votesData, curSymbol) => {
  const popoverVotesData = votesData.slice(0, 10);
  const moreCount = votesData.length - 10;

  const popoverChildren = popoverVotesData.map(v => (
    <p key={v.voter}>
      <span className="voter">{v.voter}</span>
      <span className="percent">{v.percent.toFixed(1)}</span>
      <span className="reward">
        {v.reward.toFixed(2)}
        {curSymbol}
      </span>
    </p>
  ));

  if (moreCount > 0) {
    popoverChildren.push(
      <p key="more" className="more">
        <FormattedMessage id="voters-info.n-more" values={{ n: moreCount }} />
      </p>
    );
  }

  return <div className="voters-info-popover-content">{popoverChildren}</div>;
};

class VotersInfo extends Component {
  constructor(props) {
    super(props);

    this.state = {
      modalVisible: false,
      hidePopover: false,
      popoverVisible: false
    };
  }

  popoverVisibleChanged(e) {
    this.setState({
      popoverVisible: e
    });
  }

  showModal = () => {
    this.setState({
      modalVisible: true,
      hidePopover: true
    });
  };

  handleModalCancel = () => {
    this.setState({
      modalVisible: false
    });
  };

  afterModalClosed = () => {
    this.setState({
      hidePopover: false
    });
  };

  render() {
    const { entry, children } = this.props;
    const { hidePopover, modalVisible, popoverVisible } = this.state;

    let popoverProps = {};
    let votesData = [];

    if (popoverVisible) {
      const currency = 'usd';
      const currencyRate = 1;
      const curSymbol = currencySymbol(currency);

      votesData = prepareContentVotes(entry, currencyRate);
      const popoverContent = preparePopoverContent(votesData, curSymbol);

      popoverProps = { content: popoverContent };
      if (hidePopover) {
        popoverProps.visible = false;
      }
    }

    if (entry.active_votes.length > 0) {
      const modalTableColumns = [
        {
          title: <FormattedMessage id="voters-info.author" />,
          dataIndex: 'voter',
          key: 'voter',
          width: 220,
          sorter: (a, b) => a.reputation - b.reputation,
          render: (text, record) => (
            <span>
              {text}{' '}
              <Badge
                count={record.reputation}
                style={{
                  backgroundColor: '#fff',
                  color: '#999',
                  boxShadow: '0 0 0 1px #d9d9d9 inset'
                }}
              />
            </span>
          )
        },
        {
          title: <FormattedMessage id="voters-info.reward" />,
          dataIndex: 'reward',
          key: 'reward',
          align: 'center',
          width: 150,
          defaultSortOrder: 'descend',
          sorter: (a, b) => a.reward - b.reward,
          render: text => text.toFixed(2)
        },
        {
          title: <FormattedMessage id="voters-info.percent" />,
          dataIndex: 'percent',
          key: 'percent',
          align: 'center',
          width: 150,
          sorter: (a, b) => a.percent - b.percent,
          render: text => text.toFixed(1)
        },
        {
          title: <FormattedMessage id="voters-info.time" />,
          dataIndex: 'time',
          key: 'time',
          align: 'center',
          sorter: (a, b) => a.time - b.time,
          render: text => <FormattedRelative value={text} />
        }
      ];

      return (
        <Popover
          onVisibleChange={e => {
            this.popoverVisibleChanged(e);
          }}
          {...popoverProps}
        >
          <span role="none" onClick={this.showModal}>
            {children}
          </span>

          <Modal
            visible={modalVisible}
            onCancel={this.handleModalCancel}
            afterClose={this.afterModalClosed}
            footer={false}
            width="80%"
            title={
              <FormattedMessage
                id="voters-info.modal-title"
                values={{ n: votesData.length }}
              />
            }
            centered
          >
            <Table
              dataSource={votesData}
              columns={modalTableColumns}
              scroll={{ y: 300 }}
              rowKey="voter"
            />
          </Modal>
        </Popover>
      );
    }

    // When content has no votes
    return children;
  }
}

VotersInfo.propTypes = {
  entry: PropTypes.shape({
    pending_payout_value: PropTypes.string.isRequired,
    total_payout_value: PropTypes.string.isRequired,
    curator_payout_value: PropTypes.string.isRequired,
    active_votes: PropTypes.array.isRequired
  }).isRequired,
  children: PropTypes.element.isRequired
};

export default VotersInfo;
