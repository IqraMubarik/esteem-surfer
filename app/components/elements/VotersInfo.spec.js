/* eslint-disable */
import { spy } from 'sinon';
import React from 'react';
import { prepareContentVotes, preparePopoverContent } from './VotersInfo';

const test_data_1 = {
  total_payout_value: '0.000 SBD',
  curator_payout_value: '0.000 SBD',
  pending_payout_value: '0.905 SBD',
  active_votes: [
    {
      voter: 'camille1234',
      weight: 37746,
      rshares: '79159998491',
      percent: 10000,
      reputation: '3112527265845',
      time: '2018-09-18T14:26:48'
    },
    {
      voter: 'sift666',
      weight: 4943,
      rshares: '62835135084',
      percent: 10000,
      reputation: '28708716428263',
      time: '2018-09-18T07:48:36'
    },
    {
      voter: 'kiwideb',
      weight: 39674,
      rshares: '320761392333',
      percent: 10000,
      reputation: '49978584980731',
      time: '2018-09-18T07:47:45'
    },
    {
      voter: 'mg-nz',
      weight: 127,
      rshares: 1822802530,
      percent: 10000,
      reputation: '967102003310',
      time: '2018-09-18T07:48:03'
    },
    {
      voter: 'kiwiscanfly',
      weight: 16052,
      rshares: '109182542980',
      percent: 10000,
      reputation: '29393565693170',
      time: '2018-09-18T07:52:54'
    },
    {
      voter: 'froyoempire',
      weight: 2112,
      rshares: '4428400072',
      percent: 10000,
      reputation: '1781222920987',
      time: '2018-09-19T15:15:57'
    },
    {
      voter: 'manorvillemike',
      weight: 5581,
      rshares: '11705317664',
      percent: 3300,
      reputation: '7271743051883',
      time: '2018-09-18T13:28:48'
    },
    {
      voter: 'palikari123',
      weight: 3228,
      rshares: '10830350260',
      percent: 5000,
      reputation: '5552644458720',
      time: '2018-09-18T08:02:24'
    },
    {
      voter: 'magic8ball',
      weight: 24,
      rshares: 54868820,
      percent: 3300,
      reputation: -231320336358,
      time: '2018-09-18T07:43:45'
    },
    {
      voter: 'teamnz',
      weight: 661,
      rshares: '39143612782',
      percent: 10000,
      reputation: '553862099874',
      time: '2018-09-18T07:43:45'
    },
    {
      voter: 'frot',
      weight: 2039,
      rshares: '222520508197',
      percent: 10000,
      reputation: '14611334906698',
      time: '2018-09-18T07:43:51'
    },
    {
      voter: 'rabiagilani',
      weight: 7,
      rshares: 59298464,
      percent: 4000,
      reputation: '769860848613',
      time: '2018-09-18T07:52:12'
    },
    {
      voter: 'forrestwalker',
      weight: 119,
      rshares: 552646703,
      percent: 10000,
      reputation: 6969863,
      time: '2018-09-18T07:57:18'
    }
  ]
};

describe('VotersInfo', () => {
  it('(1) prepareContentVotes', () => {
    expect(prepareContentVotes(test_data_1, 1)).toMatchSnapshot();
  });

  it('(2) preparePopoverContent', () => {
    const votesData = prepareContentVotes(test_data_1, 1);
    expect(preparePopoverContent(votesData, '$')).toMatchSnapshot();
  });
});
