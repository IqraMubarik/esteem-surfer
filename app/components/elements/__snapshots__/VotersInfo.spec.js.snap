// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VotersInfo (1) prepareVotes 1`] = `
Array [
  Object {
    "percent": 100,
    "reputation": 67,
    "reward": 0.3363498613806905,
    "rshares": "320761392333",
    "time": 2018-09-18T07:47:45.000Z,
    "voter": "kiwideb",
    "weight": 39674,
  },
  Object {
    "percent": 100,
    "reputation": 62,
    "reward": 0.23333463401580862,
    "rshares": "222520508197",
    "time": 2018-09-18T07:43:51.000Z,
    "voter": "frot",
    "weight": 2039,
  },
  Object {
    "percent": 100,
    "reputation": 65,
    "reward": 0.11448863259200961,
    "rshares": "109182542980",
    "time": 2018-09-18T07:52:54.000Z,
    "voter": "kiwiscanfly",
    "weight": 16052,
  },
  Object {
    "percent": 100,
    "reputation": 56,
    "reward": 0.08300704248004441,
    "rshares": "79159998491",
    "time": 2018-09-18T14:26:48.000Z,
    "voter": "camille1234",
    "weight": 37746,
  },
  Object {
    "percent": 100,
    "reputation": 65,
    "reward": 0.06588881792045405,
    "rshares": "62835135084",
    "time": 2018-09-18T07:48:36.000Z,
    "voter": "sift666",
    "weight": 4943,
  },
  Object {
    "percent": 100,
    "reputation": 49,
    "reward": 0.04104592712173051,
    "rshares": "39143612782",
    "time": 2018-09-18T07:43:45.000Z,
    "voter": "teamnz",
    "weight": 661,
  },
  Object {
    "percent": 33,
    "reputation": 59,
    "reward": 0.012274176592973656,
    "rshares": "11705317664",
    "time": 2018-09-18T13:28:48.000Z,
    "voter": "manorvillemike",
    "weight": 5581,
  },
  Object {
    "percent": 50,
    "reputation": 58,
    "reward": 0.011356687231465651,
    "rshares": "10830350260",
    "time": 2018-09-18T08:02:24.000Z,
    "voter": "palikari123",
    "weight": 3228,
  },
  Object {
    "percent": 100,
    "reputation": 54,
    "reward": 0.0046436129345925675,
    "rshares": "4428400072",
    "time": 2018-09-19T15:15:57.000Z,
    "voter": "froyoempire",
    "weight": 2112,
  },
  Object {
    "percent": 100,
    "reputation": 51,
    "reward": 0.0019113876948550588,
    "rshares": 1822802530,
    "time": 2018-09-18T07:48:03.000Z,
    "voter": "mg-nz",
    "weight": 127,
  },
  Object {
    "percent": 100,
    "reputation": 25,
    "reward": 0.0005795044116580298,
    "rshares": 552646703,
    "time": 2018-09-18T07:57:18.000Z,
    "voter": "forrestwalker",
    "weight": 119,
  },
  Object {
    "percent": 40,
    "reputation": 50,
    "reward": 0.000062180270516415,
    "rshares": 59298464,
    "time": 2018-09-18T07:52:12.000Z,
    "voter": "rabiagilani",
    "weight": 7,
  },
  Object {
    "percent": 33,
    "reputation": 3,
    "reward": 0.000057535353200994924,
    "rshares": 54868820,
    "time": 2018-09-18T07:43:45.000Z,
    "voter": "magic8ball",
    "weight": 24,
  },
]
`;
